File: C++ Programs/quick sort.cpp
#include <iostream>
#include<vector>
using namespace std;
int partition(vector<int> &a,int s,int e){
    int pivot=a[e];
    int i=s-1;
    for(int j=s;j<e;j++){
        if(a[j]<pivot){
            i++;
            swap(a[i],a[j]);
        }
    }
    swap(a[i+1],a[e]);
    return i+1;
}
void quick_sort(vector<int> &arr,int s,int e){
    //Base case
    if(s>=e){
        return;
    }
    //rec case
    int p=partition(arr,s,e);//index of the pivoted element
    quick_sort(arr,s,p-1);
    quick_sort(arr,p+1,e);

}
int main() {
    vector<int> arr{10,5,2,0,7,6,4};
    cout<<arr.size()<<endl;
    int n=arr.size();
    quick_sort(arr,0,n-1);
    for(int x:arr){
        cout<<x<<" ";
    }
    return 0;
    
}

File: DSA/Sort/Quick/QuickSort.java
public class QuickSort {
    public static void quickSort(int arr[], int low, int high) {
        if (low < high) {
            int pi = partition(arr, low, high);

            quickSort(arr, low, pi - 1);
            quickSort(arr, pi + 1, high);
        }
    }

    public static int partition(int arr[], int low, int high) {
        int pivot = arr[high]; // choosing last element as pivot
        int i = (low - 1);
        for (int j = low; j < high; j++) {
            if (arr[j] <= pivot) {
                i++;
                int temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
        int temp = arr[i + 1];
        arr[i + 1] = arr[high];
        arr[high] = temp;
        return i + 1;
    }

    public static void main(String args[]) {
        int arr[] = {10, 7, 8, 9, 1, 5};
        quickSort(arr, 0, arr.length - 1);

        for (int num : arr) {
            System.out.print(num + " ");
        }
        // Output: 1 5 7 8 9 10
    }
}
